sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/bucket.scala

```diff
@@ -24,12 +27,15 @@ package org.apache.spark.sql.execution.datasources
  *
  * @param numBuckets number of buckets.
  * @param bucketColumnNames the names of the columns that used to generate the bucket id.
- * @param sortColumnNames the names of the columns that used to sort data in each bucket.
+ * @param sortColumns (name, sort direction) of columns that used to sort data in each bucket.
  */
 private[sql] case class BucketSpec(
     numBuckets: Int,
     bucketColumnNames: Seq[String],
-    sortColumnNames: Seq[String])
+    sortColumns: Seq[(String, SortDirection)]) {
```

Comment:
It is great to have `SortDirection` defined explicitly. So, when we create it we know what will the order by.

However, this change implies that we can use `descending` as the order, which is actually not allowed because `UnsafeKVExternalSorter` only sort keys with ascending order (this sorter is used in `DynamicPartitionWriterContainer` when we generate bucketing files). So, I think for now, it is better to revert this change. In future, we can revisit it when we store sort direction in metastore and we can actually sort rows in a bucket file with a descending order.